python/nav/mibs/cpqpower_mib.py

```diff
@@ -0,0 +1,234 @@
+        },
+    }
+},
+{
+    'extra_columns': ['pdu3TemperatureProbeStatus'],
+    'filter': lambda x: x.get('pdu3TemperatureProbeStatus') == 'connected',
+    'sensors': {
+        'pdu3TemperatureValue': {
+            # Get from pdu3TemperatureScale to support F
+            'unit_of_measurement': Sensor.UNIT_CELSIUS,
+            'name': '{pdu3TemperatureName} Temperature',
+            'precision': 1,
+        },
+    }
+},
+{
+    'extra_columns': ['pdu3HumidityProbeStatus'],
+    'filter': lambda x: x.get('pdu3HumidityProbeStatus') == 'connected',
+    'sensors': {
+        'pdu3HumidityValue': {
+            'unit_of_measurement': Sensor.UNIT_PERCENT_RELATIVE_HUMIDITY,
+            'name': '{pdu3HumidityName} Humidity',
+            'precision': 1,
+        },
+    }
+},
+]
+
+
+class CPQPowerMib(mibretriever.MibRetriever):
+    """ Custom class for retrieveing sensors from APC UPSes."""
+    mib = get_mib('CPQPOWER-MIB')
+
+    @defer.inlineCallbacks
+    def get_all_sensors(self):
+        """Gets all the interesting sensors for this device."""
+        names = yield self._get_names()
+        result = []
+        for config in SENSORS:
+            r = yield self._get_sensors(names, **config)
+            result.extend(r)
+        defer.returnValue(result)
+
+    def _get_oid(self, column, index):
+        return self.mib['nodes'][column]['oid'] + index
+
+    @defer.inlineCallbacks
+    def _get_names(self):
+        result = {}
+        names = yield self.retrieve_columns(['pdu3Name'])
+        names = reduce_index(names)
+        for index, row in names.items():
+            name = row.get('pdu3Name')
+            if name is None:
+                name = ''
+            name = name.strip()
+            name = name.replace('\0', '')
+            result[index] = name
+        defer.returnValue(result)
+
+    @defer.inlineCallbacks
+    def _get_sensors(self, names, sensors, extra_columns=None,
+                     filter=lambda x: True):
+        if extra_columns is None:
+            extra_columns = []
+        entries = yield self.retrieve_columns(
+            extra_columns + list(sensors.keys()))
+        entries = self.translate_result(entries)
+
+        result = []
+        for index, row in iteritems(entries):
+            if not filter(row):
+                continue
+            result.extend(self._mksensors(index, row, sensors, names))
+        defer.returnValue(result)
+
+    def _mksensors(self, index, row, table, names):
+        result = []
+        for col, config in table.items():
+            if row.get(col) < 0:
+                continue
+            sensor = dict(
+                oid=self._get_oid(col, index),
+                internal_name='{}{}'.format(col, index),
+                mib=self.get_module_name()
+            )
+            for key, value in config.items():
+                if callable(value):
+                    value = value(row)
+                elif isinstance(value, text_type):
```

Comment:
It seems this will break on Python 2 if the value is unicode and not str. 

The templates are instead inserted into the database, and then ipdevinfo crashes as it tries to build a graphite request with an alias containing a format argument that doesn't exist.